### ISSUE TITLE
Option of doing optimizer.zero_grad(set_to_none=True) with accumulate_grad_batches > 1

### ISSUE THREAD

**Issue by rahumble** [PyTorchLightning/pytorch-lightning#8228]
## 🚀 Feature
An option to set gradients to None instead of zero, while also setting `accumulate_grad_batches > 1`. There might be other incompatible `Trainer` flags that I'm not aware of.

### Motivation
Training Speed improvements.

See #6534 for a previous discussion on how to do this for a simpler training situation.

### Pitch

If I have
```
trainer = pl.Trainer(accumulate_grad_batches=2, ....)
```
and
```
def optimizer_zero_grad(self, epoch, batch_idx, optimizer, optimizer_idx):
    optimizer.zero_grad(set_to_none=True)
```
I get
```pytorch_lightning.utilities.exceptions.MisconfigurationException: When overriding `LightningModule` optimizer_step or optimizer_zero_grad, `accumulate_grad_batches` in `Trainer` should be 1. It ensures optimizer_step or optimizer_zero_grad are called on every batch.```

This should work as written (or something quite close to it).

**Comment by carmocca** [comment]
Hi!

This constraint has been relaxed so it's possible to override the hook in master. A warning is printed now.

Done in #7980
